_pytest/logging.py

```diff
@@ -0,0 +1,356 @@
+        """Sets the level for capturing of logs.
+
+        By default, the level is set on the handler used to capture
+        logs. Specify a logger name to instead set the level of any
+        logger.
+        """
+
+        obj = logger and logging.getLogger(logger) or self.handler
+        obj.setLevel(level)
+
+    @contextmanager
+    def at_level(self, level, logger=None):
+        """Context manager that sets the level for capturing of logs.
+
+        By default, the level is set on the handler used to capture
+        logs. Specify a logger name to instead set the level of any
+        logger.
+        """
+        if logger is None:
+            logger = self.handler
+        else:
+            logger = logging.getLogger(logger)
+
+        orig_level = logger.level
+        logger.setLevel(level)
+        try:
+            yield
+        finally:
+            logger.setLevel(orig_level)
+
+
+@pytest.fixture
+def caplog(request):
+    """Access and control log capturing.
+
+    Captured logs are available through the following methods::
+
+    * caplog.text()          -> string containing formatted log output
+    * caplog.records()       -> list of logging.LogRecord instances
+    * caplog.record_tuples() -> list of (logger_name, level, message) tuples
+    """
+    return LogCaptureFixture(request.node)
+
+
+def get_actual_log_level(config, setting_name):
+    """Return the actual logging level."""
+    log_level = get_option_ini(config, setting_name)
+    if not log_level:
+        return
+    if isinstance(log_level, py.builtin.text):
+        log_level = log_level.upper()
+    try:
+        return int(getattr(logging, log_level, log_level))
+    except ValueError:
+        # Python logging does not recognise this as a logging level
+        raise pytest.UsageError(
+            "'{0}' is not recognized as a logging level name for "
+            "'{1}'. Please consider passing the "
+            "logging level num instead.".format(
+                log_level,
+                setting_name))
+
+
+def pytest_configure(config):
+    config.pluginmanager.register(LoggingPlugin(config), 'loggingp')
+
+
+class LoggingPlugin(object):
+    """Attaches to the logging module and captures log messages for each test.
+    """
+
+    def __init__(self, config):
+        """Creates a new plugin to capture log messages.
+
+        The formatter can be safely shared across all handlers so
+        create a single one for the entire test session here.
+        """
+        log_cli_level = get_actual_log_level(config, 'log_cli_level')
+        if log_cli_level is None:
+            # No specific CLI logging level was provided, let's check
+            # log_level for a fallback
+            log_cli_level = get_actual_log_level(config, 'log_level')
+            if log_cli_level is None:
+                # No log_level was provided, default to WARNING
+                log_cli_level = logging.WARNING
+        self.log_cli_level = log_cli_level
```

Comment:
I suggest changing `get_actual_log_level` to receive a list of options, and it returns the first non-`None`. This way this entire block becomes:

```python
self.log_cli_level = get_actual_log_level(config, 'log_cli_level', 'log_level') or logging.WARNING
```

The same is valid for `get_option_ini`.

Reply:
Nice, this for this suggestion, this works and simplifies the code a bit.